core/client/fs/src/main/java/alluxio/client/block/stream/LocalFileSecureDataReader.java

```diff
@@ -0,0 +1,176 @@
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.client.ReadType;
+import alluxio.client.file.FileSystemContext;
+import alluxio.client.file.options.InStreamOptions;
+import alluxio.conf.PropertyKey;
+import alluxio.metrics.MetricKey;
+import alluxio.metrics.MetricsSystem;
+import alluxio.network.protocol.databuffer.DataBuffer;
+import alluxio.network.protocol.databuffer.NioDataBuffer;
+import alluxio.wire.WorkerNetAddress;
+
+import alluxio.worker.block.io.LocalFileSecureBlockReader;
+import com.google.common.base.Preconditions;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.epoll.EpollEventLoopGroup;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FileDescriptor;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.file.Paths;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A secure data reader that reads packets from a local file .
+ */
+@NotThreadSafe
+public final class LocalFileSecureDataReader implements DataReader {
+  private static final Logger LOG = LoggerFactory.getLogger(LocalFileSecureDataReader.class);
+
+  /**
+   * The file reader to read a local block.
+   */
+  private final LocalFileSecureBlockReader mReader;
+  private final long mEnd;
+  private final long mChunkSize;
+  private long mPos;
+  private boolean mClosed;
+ 
+  /**
+   * Creates an instance of {@link LocalFileSecureDataReader}.
+   *
+   * @param reader    the file reader to the block path
+   * @param offset    the offset
+   * @param len       the length to read
+   * @param chunkSize the chunk size
+   */
+  private LocalFileSecureDataReader(LocalFileSecureBlockReader reader, long offset, long len, long chunkSize) {
+    mReader = reader;
+    Preconditions.checkArgument(chunkSize > 0);
+    mPos = offset;
+    mEnd = Math.min(mReader.getLength(), offset + len);
+    mChunkSize = chunkSize;
+  }
+  
+  @Override
+  public DataBuffer readChunk() throws IOException {
+    if (mPos >= mEnd) {
+      return null;
+    }
+    ByteBuffer buffer = mReader.read(mPos, Math.min(mChunkSize, mEnd - mPos));
+    DataBuffer dataBuffer = new NioDataBuffer(buffer, buffer.remaining());
+    mPos += dataBuffer.getLength();
+    MetricsSystem.counter(MetricKey.CLIENT_BYTES_READ_LOCAL.getName()).inc(dataBuffer.getLength());
+    MetricsSystem.meter(MetricKey.CLIENT_BYTES_READ_LOCAL_THROUGHPUT.getName())
+        .mark(dataBuffer.getLength());
+    return dataBuffer;
+  }
+
+  @Override
+  public long pos() {
+    return mPos;
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (mClosed) {
+      return;
+    }
+    mClosed = true;
+    mReader.decreaseUsageCount();
+  }
+
+  /**
+   * Factory class to create {@link LocalFileSecureDataReader}s.
+   */
+  @NotThreadSafe
+  public static class Factory implements DataReader.Factory {
```

Comment:
I might prefer to have a common super class for LocalFileSecureDataReader and LocalFileDataReader, then we could put the Factory in the super class.  what do you think?